src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexMatcher.cs

```diff
@@ -170,109 +200,118 @@ private SymbolicRegexMatcher(SymbolicRegexNode<TSet> rootNode, int captureCount,
+                }
+                else if (!Solver.IsEmpty(Solver.And(_builder._wordLetterForBoundariesSet, minterm)))
+                {
+                    Debug.Assert(Solver.IsEmpty(Solver.And(Solver.Not(_builder._wordLetterForBoundariesSet), minterm)));
+                    // The minterm is a subset of word letters as considered by \b and \B
+                    return CharKind.WordLetter;
+                }
+                else
+                {
+                    // All other minterms belong to the general kind
+                    return CharKind.General;
                 }
-                _asciiCharKinds = asciiCharKinds;
             }
         }
 
         /// <summary>
         /// Create a PerThreadData with the appropriate parts initialized for this matcher's pattern.
         /// </summary>
-        internal PerThreadData CreatePerThreadData() => new PerThreadData(_builder, _capsize);
+        internal PerThreadData CreatePerThreadData() => new PerThreadData(_capsize);
 
-        /// <summary>Compute the target state for the source state and input[i] character and transition to it.</summary>
-        /// <param name="builder">The associated builder.</param>
-        /// <param name="input">The input text.</param>
-        /// <param name="i">The index into <paramref name="input"/> at which the target character lives.</param>
-        /// <param name="state">The current state being transitioned from. Upon return it's the new state if the transition succeeded.</param>
+        /// <summary>Look up what is the character kind given a position ID</summary>
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        private bool TryTakeTransition<TStateHandler>(SymbolicRegexBuilder<TSet> builder, ReadOnlySpan<char> input, int i, ref CurrentState state)
-            where TStateHandler : struct, IStateHandler
+        private uint GetPositionKind(int positionId) => _positionKinds[positionId + 1];
+
+        /// <summary>
+        /// Lookup the actual minterm based on its ID. Also get its character kind, which is a general categorization of
+        /// characters used for cheaply deciding the nullability of anchors.
+        /// </summary>
+        internal TSet GetMintermFromId(int mintermId)
         {
-            int c = input[i];
+            TSet[] minterms = _minterms;
 
-            // Find the minterm, handling the special case for the last \n for states that start with a relevant anchor
-            int mintermId = c == '\n' && i == input.Length - 1 && TStateHandler.StartsWithLineAnchor(builder, ref state) ?
-                builder._minterms!.Length : // mintermId = minterms.Length represents an \n at the very end of input
-                _mintermClassifier.GetMintermID(c);
+            // A minterm ID of minterms.Length represents a \n at the very end of input, which is matched
+            // by the \Z anchor.
+            if ((uint)mintermId == (uint)minterms.Length)
```

Comment:
If you make this:
```suggestion
            if ((uint)mintermId >= (uint)minterms.Length)
```
it should eliminate the bounds check in the array indexing below.  Presumably mintermId can't be -1 here or else the below indexing would blow up anyway.